component/discovery/kubernetes_crds/kubernetes_crds.go

```diff
@@ -0,0 +1,336 @@
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to create rest config", "err", err)
+		return
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+
+	promClientset := versioned.New(clientset.RESTClient())
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to create rest client", "err", err)
+		return
+	}
+
+	fs := k8sfs.New(clientset)
+	c.cg = configGenerator{
+		config:   c.config,
+		logger:   c.logger,
+		secretfs: fs,
+	}
+
+	for _, namespace := range c.config.Namespaces {
+		factory := promop.NewSharedInformerFactoryWithOptions(promClientset,
+			5*time.Minute,
+			promop.WithNamespace(namespace),
+			promop.WithTweakListOptions(func(opts *metav1.ListOptions) {
+				if c.config.FieldSelector != "" {
+					opts.FieldSelector = c.config.FieldSelector
+				}
+				if c.config.LabelSelector != "" {
+					opts.LabelSelector = c.config.LabelSelector
+				}
+			}))
+
+		pminf := factory.Monitoring().V1().PodMonitors().Informer()
+		pminf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAddPodMonitor,
+			UpdateFunc: c.onUpdatePodMonitor,
+			DeleteFunc: c.onDeletePodMonitor,
+		})
+		pminf.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+			level.Error(c.logger).Log("msg", "pod monitor watcher error", "err", err)
+		})
+
+		sminf := factory.Monitoring().V1().ServiceMonitors().Informer()
+		sminf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAddServiceMonitor,
+			UpdateFunc: c.onUpdateServiceMonitor,
+			DeleteFunc: c.onDeleteServiceMonitor,
+		})
+		sminf.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+			level.Error(c.logger).Log("msg", "service monitor watcher error", "err", err)
+		})
+
+		probeinf := factory.Monitoring().V1().Probes().Informer()
+		probeinf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.onAddProbe,
+			UpdateFunc: c.onUpdateProbe,
+			DeleteFunc: c.onDeleteProbe,
+		})
+		probeinf.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+			level.Error(c.logger).Log("msg", "probe watcher error", "err", err)
+		})
+
+		factory.Start(ctx.Done())
+	}
+
+	level.Info(c.logger).Log("msg", "informers  started")
+
+	c.discovery = discovery.NewManager(ctx, c.logger, discovery.Name(c.opts.ID))
+	go func() {
+		err := c.discovery.Run()
+		if err != nil {
+			level.Error(c.logger).Log("msg", "discovery manager stopped", "err", err)
+		}
+	}()
+
+	flowAppendable := prometheus.NewFanout(c.config.ForwardTo, c.opts.ID, c.opts.Registerer)
+	opts := &scrape.Options{
+		HTTPClientOptions: []promCommonConfig.HTTPClientOption{
+			promCommonConfig.WithFS(fs),
+		},
+	}
+	c.scraper = scrape.NewManager(opts, c.logger, flowAppendable)
+	defer c.scraper.Stop()
+	targetSetsChan := make(chan map[string][]*targetgroup.Group)
+	go func() {
+		err := c.scraper.Run(targetSetsChan)
+		level.Info(c.logger).Log("msg", "scrape manager stopped")
+		if err != nil {
+			level.Error(c.logger).Log("msg", "scrape manager failed", "err", err)
+		}
+	}()
```

Comment:
I'm not sure all of these need to be associated with the `crdManager`. As it is now, if you change the config for the component, all scraping stops, which will interrupt the flow of metrics.

At least the scraper and appendable should be handled a layer up the goroutine tree, potentially also the discovery manager.